### PR TITLE
fix(CLAPI) fix clapi dependency duplicate insertion

### PR DESCRIPTION
**Pull request by loiclau**
## Description

When adding dependencies with CLAPI, and because there is no checks on the code nor databases, you can add several time the same child or parent.

## Type of change

- [x] Patch fixing an issue (non-breaking change)
- [ ] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [x] 19.10.x
- [x] 20.04.x
- [x] 20.10.x
- [x] 21.04.x (master)

<h2> How this pull request can be tested ? </h2>

Create a host with multiple services,
From a terminal, using CLAPI, add a dependency:
     centreon -u admin -p centreon -o DEP -a ADD -v "MyDep;MyDep;SERVICE;MyHost,Service1"
Add a child to this dependency:
     centreon -u admin -p centreon -o DEP -a ADDCHILD -v "MyDep;MyHost,Service2"
Re-execute the same command once again.

## Checklist

- [x] I have followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have made corresponding changes to the **documentation**.
- [ ] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

**Review by jeremyjaouen — DISMISSED** [review]

**Review by jeremyjaouen — DISMISSED** [review]

File: www/install/php/Update-21.04.0-beta.1.php [hunk 1]
```diff
@@ -18,3 +18,71 @@
...
+        'dependency_servicegroupChild_relation',
+        'dependency_servicegroupParent_relation'
+    ];
+    $pearDB->beginTransaction();
+
+    foreach ($tables as $table) {
+        $errorMessage = "Unable to check $tables dependencies";
+        $result = $pearDB->query(
+            "SELECT COUNT( * ) AS duplicate, host_host_id, service_service_id, dependency_dep_id
+        FROM {$tables}
```
Comment (callapa) [inline www/install/php/Update-21.04.0-beta.1.php:50]:
```suggestion
            FROM {$tables}
```

File: www/install/php/Update-21.04.0-beta.1.php [hunk 2]
```diff
@@ -18,3 +18,71 @@
...
+        GROUP BY host_host_id, dependency_dep_id, service_service_id"
```
Comment (callapa) [inline www/install/php/Update-21.04.0-beta.1.php:51]:
```suggestion
            GROUP BY host_host_id, dependency_dep_id, service_service_id"
```

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 3]
```diff
@@ -692,12 +692,18 @@ protected function addRelations($parameters, $relType = 'parent')
...
-        $sql = "INSERT INTO {$table} (dependency_dep_id, hostgroup_hg_id) VALUES (?, ?)";
         $obj = new \Centreon_Object_Host_Group($this->dependencyInjector);
         $ids = $obj->getIdByParameter($obj->getUniqueLabelField(), array($objectToInsert));
         if (!count($ids)) {
             throw new CentreonClapiException(sprintf('Could not find host group %s', $objectToInsert));
         }
+        $dataField = ['dependency_dep_id' => $depId, 'hostgroup_hg_id' => $ids[0]];
+        if ($this->isExistingDependency($table, $dataField)) {
+            throw new CentreonClapiException(
+                sprintf('Dependency between host %s already exist', $objectToInsert)
```
Comment (kduret) [inline www/class/centreon-clapi/centreonDependency.class.php:703]: `between host` and what ?

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 4]
```diff
@@ -710,12 +716,18 @@ protected function addHostgroupRelations($depId, $objectToInsert, $relType)
...
-        $sql = "INSERT INTO {$table} (dependency_dep_id, servicegroup_sg_id) VALUES (?, ?)";
         $obj = new \Centreon_Object_Service_Group($this->dependencyInjector);
         $ids = $obj->getIdByParameter($obj->getUniqueLabelField(), array($objectToInsert));
         if (!count($ids)) {
             throw new CentreonClapiException(sprintf('Could not find service group %s', $objectToInsert));
         }
+        $dataField = ['dependency_dep_id' => $depId, 'servicegroup_sg_id' => $ids[0]];
+        if ($this->isExistingDependency($table, $dataField)) {
+            throw new CentreonClapiException(
+                sprintf('Dependency between host %s already exist', $objectToInsert)
```
Comment (kduret) [inline www/class/centreon-clapi/centreonDependency.class.php:727]: seems not to be a host dependency...

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 5]
```diff
@@ -747,30 +764,51 @@ protected function addMetaRelations($depId, $objectToInsert, $relType)
...
-            $sql = "INSERT INTO dependency_hostParent_relation (dependency_dep_id, host_host_id) VALUES (?, ?)";
             $hostObj = new \Centreon_Object_Host($this->dependencyInjector);
             $hostIds = $hostObj->getIdByParameter($hostObj->getUniqueLabelField(), array($objectToInsert));
             if (!count($hostIds)) {
                 throw new CentreonClapiException(sprintf('Could not find host %s', $objectToInsert));
             }
+            $dataField = ['dependency_dep_id' => $depId, 'host_host_id' => $hostIds[0]];
+            if ($this->isExistingDependency('dependency_hostParent_relation', $dataField)) {
+                throw new CentreonClapiException(
+                    sprintf('Dependency between host %s already exist', $objectToInsert)
```
Comment (kduret) [inline www/class/centreon-clapi/centreonDependency.class.php:775]: `between host` and what ?

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 6]
```diff
@@ -747,30 +764,51 @@ protected function addMetaRelations($depId, $objectToInsert, $relType)
...
-            $sql = "INSERT INTO dependency_hostChild_relation (dependency_dep_id, host_host_id) VALUES (?, ?)";
             $hostObj = new \Centreon_Object_Host($this->dependencyInjector);
             $hostIds = $hostObj->getIdByParameter($hostObj->getUniqueLabelField(), array($objectToInsert));
             if (!count($hostIds)) {
                 throw new CentreonClapiException(sprintf('Could not find host %s', $objectToInsert));
             }
+            $dataField = ['dependency_dep_id' => $depId, 'host_host_id' => $hostIds[0]];
+            if ($this->isExistingDependency('dependency_hostChild_relation', $dataField)) {
+                throw new CentreonClapiException(
+                    sprintf('Dependency between host %s already exist', $objectToInsert)
```
Comment (kduret) [inline www/class/centreon-clapi/centreonDependency.class.php:808]: `between host` and what ?

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 7]
```diff
@@ -862,6 +926,27 @@ protected function deleteRelations($parameters, $relType = 'parent')
...
+        foreach ($dataField as $field => $value) {
+            $sql .= " {$field} = {$value} AND";
+        }
+        $sql = rtrim($sql, "AND");
+        $res = $this->db->query($sql);
+        $row = $res->fetch();
+        if (!empty($row['dependency_dep_id'])) {
+            return true;
+        }
+        return false;
```
Comment (kduret) [inline www/class/centreon-clapi/centreonDependency.class.php:947]:
```suggestion
        return !empty($row['dependency_dep_id']);
```

File: www/install/php/Update-20.10.4.php [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+ *
+ * For more information : contact@centreon.com
+ *
+ */
+
+include_once __DIR__ . "/../../class/centreonLog.class.php";
+$centreonLog = new CentreonLog();
+
+//error specific content
+$versionOfTheUpgrade = 'UPGRADE - 21.10.4 : ';
```
Comment (kduret) [inline www/install/php/Update-20.10.4.php:26]:
```suggestion
$versionOfTheUpgrade = 'UPGRADE - 20.10.4 : ';
```

**Review by kduret — CHANGES_REQUESTED** [review]

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 9]
```diff
@@ -692,12 +706,19 @@ protected function addRelations($parameters, $relType = 'parent')
...
         $obj = new \Centreon_Object_Host_Group($this->dependencyInjector);
         $ids = $obj->getIdByParameter($obj->getUniqueLabelField(), array($objectToInsert));
         if (!count($ids)) {
             throw new CentreonClapiException(sprintf('Could not find host group %s', $objectToInsert));
         }
+        $dataField = ['dependency_dep_id' => $depId, 'hostgroup_hg_id' => $ids[0]];
+        if ($this->isExistingDependency($table, $dataField)) {
+            $depName = $this->getDependencyName($depId);
+            throw new CentreonClapiException(
+                sprintf('Hostgroup %s already link to the dependency %s', $objectToInsert, $depName)
```
Comment (jeremyjaouen) [inline www/class/centreon-clapi/centreonDependency.class.php:718]:
```suggestion
                sprintf('Hostgroup %s already link to the dependency %s', $objectToInsert, $this->getDependencyName($depId))
```

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 10]
```diff
@@ -692,12 +706,19 @@ protected function addRelations($parameters, $relType = 'parent')
...
+            );
+        }
+        $sql = "INSERT INTO {$table} (dependency_dep_id, hostgroup_hg_id) VALUES (?, ?)";
         $this->db->query($sql, array($depId, $ids[0]));
```
Comment (jeremyjaouen) [inline www/class/centreon-clapi/centreonDependency.class.php:722]: why not using a prepare / bind / execute statement ?

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 11]
```diff
@@ -710,12 +731,19 @@ protected function addHostgroupRelations($depId, $objectToInsert, $relType)
...
         $obj = new \Centreon_Object_Service_Group($this->dependencyInjector);
         $ids = $obj->getIdByParameter($obj->getUniqueLabelField(), array($objectToInsert));
         if (!count($ids)) {
             throw new CentreonClapiException(sprintf('Could not find service group %s', $objectToInsert));
         }
+        $dataField = ['dependency_dep_id' => $depId, 'servicegroup_sg_id' => $ids[0]];
+        if ($this->isExistingDependency($table, $dataField)) {
+            $depName = $this->getDependencyName($depId);
+            throw new CentreonClapiException(
+                sprintf('Servicegroup %s already link to the dependency %s', $objectToInsert, $depName)
```
Comment (jeremyjaouen) [inline www/class/centreon-clapi/centreonDependency.class.php:743]:
```suggestion
                sprintf('Servicegroup %s already link to the dependency %s', $objectToInsert, $this->getDependencyName($depId))
```

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 12]
```diff
@@ -710,12 +731,19 @@ protected function addHostgroupRelations($depId, $objectToInsert, $relType)
...
+            );
+        }
+        $sql = "INSERT INTO {$table} (dependency_dep_id, servicegroup_sg_id) VALUES (?, ?)";
         $this->db->query($sql, array($depId, $ids[0]));
```
Comment (jeremyjaouen) [inline www/class/centreon-clapi/centreonDependency.class.php:747]: why not using a prepare / bind / execute statement ?

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 13]
```diff
@@ -710,12 +731,19 @@ protected function addHostgroupRelations($depId, $objectToInsert, $relType)
...
         $table = "dependency_servicegroup" . ucfirst($relType) . "_relation";
-        $sql = "INSERT INTO {$table} (dependency_dep_id, servicegroup_sg_id) VALUES (?, ?)";
         $obj = new \Centreon_Object_Service_Group($this->dependencyInjector);
         $ids = $obj->getIdByParameter($obj->getUniqueLabelField(), array($objectToInsert));
         if (!count($ids)) {
             throw new CentreonClapiException(sprintf('Could not find service group %s', $objectToInsert));
         }
+        $dataField = ['dependency_dep_id' => $depId, 'servicegroup_sg_id' => $ids[0]];
+        if ($this->isExistingDependency($table, $dataField)) {
+            $depName = $this->getDependencyName($depId);
```
Comment (jeremyjaouen) [inline www/class/centreon-clapi/centreonDependency.class.php:741]:
```suggestion

```

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 14]
```diff
@@ -692,12 +706,19 @@ protected function addRelations($parameters, $relType = 'parent')
...
         $table = "dependency_hostgroup" . ucfirst($relType) . "_relation";
-        $sql = "INSERT INTO {$table} (dependency_dep_id, hostgroup_hg_id) VALUES (?, ?)";
         $obj = new \Centreon_Object_Host_Group($this->dependencyInjector);
         $ids = $obj->getIdByParameter($obj->getUniqueLabelField(), array($objectToInsert));
         if (!count($ids)) {
             throw new CentreonClapiException(sprintf('Could not find host group %s', $objectToInsert));
         }
+        $dataField = ['dependency_dep_id' => $depId, 'hostgroup_hg_id' => $ids[0]];
+        if ($this->isExistingDependency($table, $dataField)) {
+            $depName = $this->getDependencyName($depId);
```
Comment (jeremyjaouen) [inline www/class/centreon-clapi/centreonDependency.class.php:716]:
```suggestion

```

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 15]
```diff
@@ -728,13 +756,19 @@ protected function addServicegroupRelations($depId, $objectToInsert, $relType)
...
-
-        $sql = "INSERT INTO {$table} (dependency_dep_id, meta_service_meta_id) VALUES (?, ?)";
         $obj = new \Centreon_Object_Meta_Service($this->dependencyInjector);
         $ids = $obj->getIdByParameter($obj->getUniqueLabelField(), array($objectToInsert));
         if (!count($ids)) {
             throw new CentreonClapiException(sprintf('Could not find meta service %s', $objectToInsert));
         }
+        $dataField = ['dependency_dep_id' => $depId, 'meta_service_meta_id' => $ids[0]];
+        if ($this->isExistingDependency($table, $dataField)) {
+            $depName = $this->getDependencyName($depId);
```
Comment (jeremyjaouen) [inline www/class/centreon-clapi/centreonDependency.class.php:766]:
```suggestion

```

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 16]
```diff
@@ -728,13 +756,19 @@ protected function addServicegroupRelations($depId, $objectToInsert, $relType)
...
+            throw new CentreonClapiException(
+                sprintf('Meta %s already already link to the dependency %s', $objectToInsert, $depName)
```
Comment (jeremyjaouen) [inline www/class/centreon-clapi/centreonDependency.class.php:768]:
```suggestion
                sprintf('Meta %s already already link to the dependency %s', $objectToInsert, $this->getDependencyName($depId))
```

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 17]
```diff
@@ -728,13 +756,19 @@ protected function addServicegroupRelations($depId, $objectToInsert, $relType)
...
+            );
+        }
+        $sql = "INSERT INTO {$table} (dependency_dep_id, meta_service_meta_id) VALUES (?, ?)";
         $this->db->query($sql, array($depId, $ids[0]));
```
Comment (jeremyjaouen) [inline www/class/centreon-clapi/centreonDependency.class.php:772]: why not using a prepare / bind / execute statement ?

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 18]
```diff
@@ -747,30 +781,54 @@ protected function addMetaRelations($depId, $objectToInsert, $relType)
...
         if ($relType == 'parent') {
-            $sql = "INSERT INTO dependency_hostParent_relation (dependency_dep_id, host_host_id) VALUES (?, ?)";
             $hostObj = new \Centreon_Object_Host($this->dependencyInjector);
             $hostIds = $hostObj->getIdByParameter($hostObj->getUniqueLabelField(), array($objectToInsert));
             if (!count($hostIds)) {
                 throw new CentreonClapiException(sprintf('Could not find host %s', $objectToInsert));
             }
+            $dataField = ['dependency_dep_id' => $depId, 'host_host_id' => $hostIds[0]];
+            if ($this->isExistingDependency('dependency_hostParent_relation', $dataField)) {
+                $depName = $this->getDependencyName($depId);
```
Comment (jeremyjaouen) [inline www/class/centreon-clapi/centreonDependency.class.php:791]:
```suggestion

```

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 19]
```diff
@@ -747,30 +781,54 @@ protected function addMetaRelations($depId, $objectToInsert, $relType)
...
+                throw new CentreonClapiException(
+                    sprintf('Host %s already link to the dependency %s', $objectToInsert, $depName)
```
Comment (jeremyjaouen) [inline www/class/centreon-clapi/centreonDependency.class.php:793]:
```suggestion
                    sprintf('Host %s already link to the dependency %s', $objectToInsert, $this->getDependencyName($depId))
```

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 20]
```diff
@@ -747,30 +781,54 @@ protected function addMetaRelations($depId, $objectToInsert, $relType)
...
             if (!count($idTab)) {
                 throw new CentreonClapiException(sprintf('Could not find service %s on host %s', $service, $host));
             }
+            $dataField = [
+                'dependency_dep_id' => $depId,
+                'host_host_id' => $idTab[0],
+                'service_service_id' => $idTab[1]
+            ];
+            if ($this->isExistingDependency('dependency_serviceChild_relation', $dataField)) {
+                $depName = $this->getDependencyName($depId);
```
Comment (jeremyjaouen) [inline www/class/centreon-clapi/centreonDependency.class.php:810]:
```suggestion

```

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 21]
```diff
@@ -747,30 +781,54 @@ protected function addMetaRelations($depId, $objectToInsert, $relType)
...
+                throw new CentreonClapiException(
+                    sprintf('Dependency between service %s and host %s already exist on %s', $service, $host, $depName)
```
Comment (jeremyjaouen) [inline www/class/centreon-clapi/centreonDependency.class.php:812]:
```suggestion
                    sprintf('Dependency between service %s and host %s already exist on %s', $service, $host, $this->getDependencyName($depId))
```

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 22]
```diff
@@ -747,30 +781,54 @@ protected function addMetaRelations($depId, $objectToInsert, $relType)
...
         } elseif ($relType == 'child') { // host child
-            $sql = "INSERT INTO dependency_hostChild_relation (dependency_dep_id, host_host_id) VALUES (?, ?)";
             $hostObj = new \Centreon_Object_Host($this->dependencyInjector);
             $hostIds = $hostObj->getIdByParameter($hostObj->getUniqueLabelField(), array($objectToInsert));
             if (!count($hostIds)) {
                 throw new CentreonClapiException(sprintf('Could not find host %s', $objectToInsert));
             }
+            $dataField = ['dependency_dep_id' => $depId, 'host_host_id' => $hostIds[0]];
+            if ($this->isExistingDependency('dependency_hostChild_relation', $dataField)) {
+                $depName = $this->getDependencyName($depId);
```
Comment (jeremyjaouen) [inline www/class/centreon-clapi/centreonDependency.class.php:826]:
```suggestion

```

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 23]
```diff
@@ -747,30 +781,54 @@ protected function addMetaRelations($depId, $objectToInsert, $relType)
...
+                throw new CentreonClapiException(
+                    sprintf('Host %s already link to the dependency %s', $objectToInsert, $depName)
```
Comment (jeremyjaouen) [inline www/class/centreon-clapi/centreonDependency.class.php:828]:
```suggestion
                    sprintf('Host %s already link to the dependency %s', $objectToInsert, $this->getDependencyName($depId))
```

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 24]
```diff
@@ -747,30 +781,54 @@ protected function addMetaRelations($depId, $objectToInsert, $relType)
...
+                );
+            }
+            $sql = "INSERT INTO dependency_hostChild_relation (dependency_dep_id, host_host_id) VALUES (?, ?)";
             $params = array($depId, $hostIds[0]);
         }
         $this->db->query($sql, $params);
```
Comment (jeremyjaouen) [inline www/class/centreon-clapi/centreonDependency.class.php:834]: why not using a prepare / bind / execute statement ? (for other queries into the if elseif too)

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 25]
```diff
@@ -796,24 +851,56 @@ protected function addServiceRelations($depId, $objectToInsert, $relType)
             if (!count($idTab)) {
                 throw new CentreonClapiException(sprintf('Could not find service %s on host %s', $service, $host));
             }
+            $dataField = [
+                'dependency_dep_id' => $depId,
+                'host_host_id' => $idTab[0],
+                'service_service_id' => $idTab[1]
+            ];
+            if ($this->isExistingDependency('dependency_serviceParent_relation', $dataField)) {
+                $depName = $this->getDependencyName($depId);
```
Comment (jeremyjaouen) [inline www/class/centreon-clapi/centreonDependency.class.php:860]:
```suggestion

```

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 26]
```diff
@@ -796,24 +851,56 @@ protected function addServiceRelations($depId, $objectToInsert, $relType)
...
+                throw new CentreonClapiException(
+                    sprintf('Dependency between service %s and host %s already exist on %s', $service, $host, $depName)
```
Comment (jeremyjaouen) [inline www/class/centreon-clapi/centreonDependency.class.php:862]:
```suggestion
                    sprintf('Dependency between service %s and host %s already exist on %s', $service, $host, $                $depName = $this->getDependencyName($depId))
```

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 27]
```diff
@@ -796,24 +851,56 @@ protected function addServiceRelations($depId, $objectToInsert, $relType)
...
+                    sprintf('Could not find service %s on host %s', $depId, $service, $host)
+                );
+            }
+            $dataField = [
+                'dependency_dep_id' => $depId,
+                'host_host_id' => $idTab[0],
+                'service_service_id' => $idTab[1]
+            ];
+            if ($this->isExistingDependency('dependency_serviceChild_relation', $dataField)) {
+                $depName = $this->getDependencyName($depId);
```
Comment (jeremyjaouen) [inline www/class/centreon-clapi/centreonDependency.class.php:882]:
```suggestion

```

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 28]
```diff
@@ -796,24 +851,56 @@ protected function addServiceRelations($depId, $objectToInsert, $relType)
...
+                throw new CentreonClapiException(
+                    sprintf('Dependency between service %s and host %s already exist on %s', $service, $host, $depName)
```
Comment (jeremyjaouen) [inline www/class/centreon-clapi/centreonDependency.class.php:884]:
```suggestion
                    sprintf('Dependency between service %s and host %s already exist on %s', $service, $host, $this->getDependencyName($depId))
```

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 29]
```diff
@@ -796,24 +851,56 @@ protected function addServiceRelations($depId, $objectToInsert, $relType)
...
-            $sql = "INSERT INTO dependency_hostChild_relation (dependency_dep_id, host_host_id)
-                VALUES (?, ?)";
             $hostObj = new \Centreon_Object_Host($this->dependencyInjector);
             $hostIds = $hostObj->getIdByParameter($hostObj->getUniqueLabelField(), array($objectToInsert));
             if (!count($hostIds)) {
                 throw new CentreonClapiException(sprintf('Could not find host %s', $objectToInsert));
             }
+            $dataField = ['dependency_dep_id' => $depId, 'host_host_id' => $hostIds[0]];
+            if ($this->isExistingDependency('dependency_serviceChild_relation', $dataField)) {
+                $depName = $this->getDependencyName($depId);
```
Comment (jeremyjaouen) [inline www/class/centreon-clapi/centreonDependency.class.php:898]:
```suggestion

```

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 30]
```diff
@@ -796,24 +851,56 @@ protected function addServiceRelations($depId, $objectToInsert, $relType)
...
+                throw new CentreonClapiException(
+                    sprintf('Service %s already link to the dependency %s', $objectToInsert, $depName)
```
Comment (jeremyjaouen) [inline www/class/centreon-clapi/centreonDependency.class.php:900]:
```suggestion
                    sprintf('Service %s already link to the dependency %s', $objectToInsert, $this->getDependencyName($depId))
```

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 31]
```diff
@@ -796,24 +851,56 @@ protected function addServiceRelations($depId, $objectToInsert, $relType)
...
+                );
+            }
+            $sql = "INSERT INTO dependency_hostChild_relation (dependency_dep_id, host_host_id) VALUES (?, ?)";
             $params = array($depId, $hostIds[0]);
         }
         $this->db->query($sql, $params);
```
Comment (jeremyjaouen) [inline www/class/centreon-clapi/centreonDependency.class.php:906]: why not using a prepare / bind / execute statement ? (for other queries into the if elseif too)

**Review by jeremyjaouen — CHANGES_REQUESTED** [review]

**Review by jeremyjaouen — DISMISSED** [review]

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 32]
```diff
@@ -692,12 +709,22 @@ protected function addRelations($parameters, $relType = 'parent')
...
         $obj = new \Centreon_Object_Host_Group($this->dependencyInjector);
         $ids = $obj->getIdByParameter($obj->getUniqueLabelField(), array($objectToInsert));
         if (!count($ids)) {
             throw new CentreonClapiException(sprintf('Could not find host group %s', $objectToInsert));
         }
+        $dataField = ['dependency_dep_id' => $depId, 'hostgroup_hg_id' => $ids[0]];
+        if ($this->isExistingDependency($table, $dataField)) {
+            throw new CentreonClapiException(
+                sprintf(
+                    'Hostgroup %s already link to the dependency %s',
```
Comment (kduret) [inline www/class/centreon-clapi/centreonDependency.class.php:721]:
```suggestion
                    'Hostgroup %s is already linked to the dependency %s',
```

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 33]
```diff
@@ -710,12 +737,22 @@ protected function addHostgroupRelations($depId, $objectToInsert, $relType)
...
         $obj = new \Centreon_Object_Service_Group($this->dependencyInjector);
         $ids = $obj->getIdByParameter($obj->getUniqueLabelField(), array($objectToInsert));
         if (!count($ids)) {
             throw new CentreonClapiException(sprintf('Could not find service group %s', $objectToInsert));
         }
+        $dataField = ['dependency_dep_id' => $depId, 'servicegroup_sg_id' => $ids[0]];
+        if ($this->isExistingDependency($table, $dataField)) {
+            throw new CentreonClapiException(
+                sprintf(
+                    'Servicegroup %s already link to the dependency %s',
```
Comment (kduret) [inline www/class/centreon-clapi/centreonDependency.class.php:749]:
```suggestion
                    'Servicegroup %s is already linked to the dependency %s',
```

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 34]
```diff
@@ -728,13 +765,22 @@ protected function addServicegroupRelations($depId, $objectToInsert, $relType)
...
         $obj = new \Centreon_Object_Meta_Service($this->dependencyInjector);
         $ids = $obj->getIdByParameter($obj->getUniqueLabelField(), array($objectToInsert));
         if (!count($ids)) {
             throw new CentreonClapiException(sprintf('Could not find meta service %s', $objectToInsert));
         }
+        $dataField = ['dependency_dep_id' => $depId, 'meta_service_meta_id' => $ids[0]];
+        if ($this->isExistingDependency($table, $dataField)) {
+            throw new CentreonClapiException(
+                sprintf(
+                    'Meta %s already already link to the dependency %s',
```
Comment (kduret) [inline www/class/centreon-clapi/centreonDependency.class.php:777]:
```suggestion
                    'Meta %s already is already linked to the dependency %s',
```

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 35]
```diff
@@ -747,30 +793,64 @@ protected function addMetaRelations($depId, $objectToInsert, $relType)
...
             $hostObj = new \Centreon_Object_Host($this->dependencyInjector);
             $hostIds = $hostObj->getIdByParameter($hostObj->getUniqueLabelField(), array($objectToInsert));
             if (!count($hostIds)) {
                 throw new CentreonClapiException(sprintf('Could not find host %s', $objectToInsert));
             }
+            $dataField = ['dependency_dep_id' => $depId, 'host_host_id' => $hostIds[0]];
+            if ($this->isExistingDependency('dependency_hostParent_relation', $dataField)) {
+                throw new CentreonClapiException(
+                    sprintf(
+                        'Host %s already link to the dependency %s',
```
Comment (kduret) [inline www/class/centreon-clapi/centreonDependency.class.php:805]:
```suggestion
                        'Host %s is already linked to the dependency %s',
```

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 36]
```diff
@@ -747,30 +793,64 @@ protected function addMetaRelations($depId, $objectToInsert, $relType)
...
             }
+            $dataField = [
+                'dependency_dep_id' => $depId,
+                'host_host_id' => $idTab[0],
+                'service_service_id' => $idTab[1]
+            ];
+            if ($this->isExistingDependency('dependency_serviceChild_relation', $dataField)) {
+                throw new CentreonClapiException(
+                    sprintf(
+                        'Dependency between service %s and host %s already exist on %s',
```
Comment (kduret) [inline www/class/centreon-clapi/centreonDependency.class.php:827]:
```suggestion
                        'Dependency between service %s and host %s already exists on %s',
```

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 37]
```diff
@@ -747,30 +793,64 @@ protected function addMetaRelations($depId, $objectToInsert, $relType)
...
             $hostObj = new \Centreon_Object_Host($this->dependencyInjector);
             $hostIds = $hostObj->getIdByParameter($hostObj->getUniqueLabelField(), array($objectToInsert));
             if (!count($hostIds)) {
                 throw new CentreonClapiException(sprintf('Could not find host %s', $objectToInsert));
             }
+            $dataField = ['dependency_dep_id' => $depId, 'host_host_id' => $hostIds[0]];
+            if ($this->isExistingDependency('dependency_hostChild_relation', $dataField)) {
+                throw new CentreonClapiException(
+                    sprintf(
+                        'Host %s already link to the dependency %s',
```
Comment (kduret) [inline www/class/centreon-clapi/centreonDependency.class.php:847]:
```suggestion
                        'Host %s is already linked to the dependency %s',
```

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 38]
```diff
@@ -796,24 +873,67 @@ protected function addServiceRelations($depId, $objectToInsert, $relType)
...
             }
+            $dataField = [
+                'dependency_dep_id' => $depId,
+                'host_host_id' => $idTab[0],
+                'service_service_id' => $idTab[1]
+            ];
+            if ($this->isExistingDependency('dependency_serviceParent_relation', $dataField)) {
+                throw new CentreonClapiException(
+                    sprintf(
+                        'Dependency between service %s and host %s already exist on %s',
```
Comment (kduret) [inline www/class/centreon-clapi/centreonDependency.class.php:884]:
```suggestion
                        'Dependency between service %s and host %s already exists on %s',
```

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 39]
```diff
@@ -796,24 +873,67 @@ protected function addServiceRelations($depId, $objectToInsert, $relType)
...
+            }
+            $dataField = [
+                'dependency_dep_id' => $depId,
+                'host_host_id' => $idTab[0],
+                'service_service_id' => $idTab[1]
+            ];
+            if ($this->isExistingDependency('dependency_serviceChild_relation', $dataField)) {
+                throw new CentreonClapiException(
+                    sprintf(
+                        'Dependency between service %s and host %s already exist on %s',
```
Comment (kduret) [inline www/class/centreon-clapi/centreonDependency.class.php:910]:
```suggestion
                        'Dependency between service %s and host %s already exists on %s',
```

File: www/class/centreon-clapi/centreonDependency.class.php [hunk 40]
```diff
@@ -796,24 +873,67 @@ protected function addServiceRelations($depId, $objectToInsert, $relType)
...
             $hostObj = new \Centreon_Object_Host($this->dependencyInjector);
             $hostIds = $hostObj->getIdByParameter($hostObj->getUniqueLabelField(), array($objectToInsert));
             if (!count($hostIds)) {
                 throw new CentreonClapiException(sprintf('Could not find host %s', $objectToInsert));
             }
+            $dataField = ['dependency_dep_id' => $depId, 'host_host_id' => $hostIds[0]];
+            if ($this->isExistingDependency('dependency_serviceChild_relation', $dataField)) {
+                throw new CentreonClapiException(
+                    sprintf(
+                        'Service %s already link to the dependency %s',
```
Comment (kduret) [inline www/class/centreon-clapi/centreonDependency.class.php:930]:
```suggestion
                        'Service %s is already linked to the dependency %s',
```

**Review by jeremyjaouen — DISMISSED** [review]

**Review by kduret — DISMISSED** [review]

File: www/class/centreon-clapi/centreonDependency.class.php (hunk 10 above)
Comment (callapa) [reply www/class/centreon-clapi/centreonDependency.class.php:728]: If you are not using the prepare request, please cast values for more security.
`$this->db->query($sql, array((int) $depId, (int) $ids[0]));`

**Review by callapa — CHANGES_REQUESTED** [review]

**Review by jeremyjaouen — APPROVED** [review]

**Review by kduret — APPROVED** [review]

**Review by callapa — APPROVED** [review]

**Pull request merged**